data/state/accountsDB.go

```diff
@@ -104,46 +109,150 @@ func (adb *AccountsDB) SaveAccount(account AccountHandler) error {
 	return adb.saveAccountToTrie(account)
 }
 
-func (adb *AccountsDB) saveCode(accountHandler baseAccountHandler) error {
-	// TODO: enable code pruning
+func (adb *AccountsDB) getOldCodeHash(oldAcc AccountHandler) ([]byte, error) {
+	if check.IfNil(oldAcc) {
+		return nil, nil
+	}
 
-	code := accountHandler.GetCode()
-	if len(code) == 0 {
-		return nil
+	oldAccount, ok := oldAcc.(baseAccountHandler)
+	if !ok {
+		return nil, ErrWrongTypeAssertion
 	}
 
-	codeHash := adb.hasher.Compute(string(code))
-	currentCodeHash := accountHandler.GetCodeHash()
-	noUpdateNeeded := bytes.Equal(codeHash, currentCodeHash)
-	if noUpdateNeeded {
+	return oldAccount.GetCodeHash(), nil
+}
+
+func (adb *AccountsDB) saveCode(accountHandler baseAccountHandler, oldAcc AccountHandler) error {
+	// TODO when state splitting is implemented, check how the code should be copied in different shards
+	oldCodeHash, err := adb.getOldCodeHash(oldAcc)
+	if err != nil {
+		return err
+	}
+
+	newCode := accountHandler.GetCode()
+	var newCodeHash []byte
+	if len(newCode) != 0 {
+		newCodeHash = adb.hasher.Compute(string(newCode))
+	}
+
+	if bytes.Equal(oldCodeHash, newCodeHash) {
```

Comment:
saveCode should be called only there is a new code. if there is something in old code - delete account / code should be called. If a code is delete the whole account is deleted.

Reply:
This also sets the codeHash for the account. If a code is updated, isn't the same account used? In this case, there is something in both oldCode and newCode.